pgrx-pg-sys/build/clang.rs

```diff
@@ -0,0 +1,129 @@
+        }
+    };
+
+    // If Postgres is configured --with-llvm, then it may have recorded a CLANG to use
+    // Ask if there's a clang at the path that Postgres would use for JIT purposes.
+    // Unfortunately, the responses from clang-sys include clangs from far-off paths,
+    // so we can only use clangs that match bindgen's libclang major version.
+    if let Some(ClangSys { path, version: Some(v), c_search_paths, .. }) =
+        ClangSys::find(preferred_clang, &[])
+    {
+        if Some(&*path) == preferred_clang && v.Major as u32 == clang_major {
+            return (false, c_search_paths.unwrap_or_default());
+        }
+    }
+
+    // Oh no, still here?
+    // Let's go behind bindgen's back to get libclang's path
+    let libclang_path =
+        clang_sys::get_library().expect("libclang should have been loaded?").path().to_owned();
+    eprintln!("found libclang at {}", libclang_path.display());
+    // libclang will probably be in a dynamic library directory,
+    // which means it will probably be adjacent to its headers, e.g.
+    // - "/usr/lib/libclang-${CLANG_MAJOR}.so.${CLANG_MAJOR}.${CLANG_MINOR}"
+    // - "/usr/lib/clang/${CLANG_MAJOR}/include"
+    let clang_major_fmt = clang_major.to_string();
+    let mut paths = vec![];
+    // by adjacent, that does not mean it is always immediately so, e.g.
+    // - "/usr/lib/x86_64-linux-gnu/libclang-${CLANG_MAJOR}.so.${CLANG_MAJOR}.${CLANG_MINOR}.${CLANG_SUBMINOR}"
+    // - "/usr/lib/clang/${CLANG_MAJOR}/include"
+    // or
+    // - "/usr/lib64/libclang-${CLANG_MAJOR}.so.${CLANG_MAJOR}.${CLANG_MINOR}.${CLANG_SUBMINOR}"
+    // - "/usr/lib/clang/${CLANG_MAJOR}/include"
+    // so, crawl back up the ancestral tree
+    for ancestor in libclang_path.ancestors() {
+        paths = WalkDir::new(ancestor)
+            .min_depth(1)
+            .max_depth(6)
+            .sort_by_file_name()
+            .into_iter()
+            // On Unix-y systems this will be like "/usr/lib/clang/$CLANG_MAJOR/include"
+            // so don't even descend if the directory doesn't have one of those parts
+            .filter_entry(|entry| {
+                !is_hidden(entry) && {
+                    entry_contains(entry, "clang")
+                        || entry_contains(entry, "include")
+                        || entry_contains(entry, &*clang_major_fmt)
+                        // we always want to descend from a lib dir, but only one step
+                        // as we don't really want to search all of /usr/lib's subdirs
+                        || os_str_contains(entry.file_name(), "lib")
+                }
+            })
+            .filter_map(|e| e.ok()) // be discreet
+            // We now need something that looks like it actually satisfies all our constraints
+            .filter(|entry| {
+                entry_contains(entry, &*clang_major_fmt)
+                    && entry_contains(entry, "clang")
+                    && entry_contains(entry, "include")
+            })
```

Comment:
The fast-and-loose way this may seem to be written is to account for possible differences like
- /usr/lib/clang-15/include
- /usr/lib/clang/15/include

It may seem unnecessary, but we don't necessarily _want_ to have a hard dependency on there being a clang executable. This is because it is possible for libclang to be *packaged without its executable* and this is a configuration that is actually shipped by real distros.

You may think that seems... bad... but it's less goofy than the configurations that package libclang separately from its headers!